### PR TITLE
Support WPA/WPA2

### PR DESCRIPTION
**Pull request by rickywu0421**
Originally, vwifi doesn't support WPA/WPA2 and can only perform hostapd without the cipher suite and key management. This commit let vwifi support WPA/WPA2 with the cipher suite TKIP/CCMP and the legacy WEP40 and WEP104, so we can run hostapd on the HostAP mode interface and wpa_supplicant on the station mode interface with WPA or WPA2. Right now we only test the WPA-PSK AKM (Authentication and Key Management) suite. To support WPA-EAP, several modifications may be needed.

This commit mainly focuses on two aspects:

1. Beacon IE: once startup hostapd on the AP, it will pass the beacon frame to the driver via `cfg80211_ops->start_ap()` or `cfg80211_ops->change_beacon()`. The beacon frame may contain several IEs like SSID IE, WPA/RSN IE, etc. We store the beacon frame IEs into the AP's vif so that when STA scan the AP it will inform cfg80211 with the IEs owned by the AP (for the latter used by wpa_supplicant).
2. Connection routine: when AP finished the connection with an STA, it call `cfg80211_new_sta()` to notify the hostapd about the success of the association. If we want hostapd to start the process of the 4-way handshake with the STA, we need to pass `cfg80211_new_sta()` with the association request with the WPA/RSN IE, so that hostapd knows what chipher suite and AKM suite should be used. Right now, we fake the association request IE by the AP's beacon frame IE. It seems safe because hostapd will ignore the unnecessary parts and focus on WPA/RSN IE only.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -153,173 +167,115 @@ Wiphy phy13
...
-	signal:  	-33 dBm
-	current time:	1655310275763 ms
-```
+The "managed mode" in the Supported interface modes is identical to station mode.
 
 ### Creating Network Namespaces
+In our testing environment, we have two network interfaces: `owl0` in HostAP mode and `owl1` in station mode. Thus, we need only one network namespace in order to isolate them.
 
-Next, create three network namespaces using the following commands:
+Create a network namespace using the following commands:
```
Comment (jserv) [inline README.md:179]: a network namespace?

Comment (rickywu0421) [reply README.md:179]: Since I have simplify the testing environment to be only one STA and one HostAP, so we just need one network namespace (one remains in `init_net` is fine). Do you think we should remain the two STA and one HostAP testing environment?

Comment (jserv) [reply README.md:179]: I expect the minimal changes against the existing code when a new feature comes in.

File: vwifi.c [hunk 2]
```diff
@@ -91,12 +106,16 @@ struct owl_vif {
...
             bool ap_enabled;
+            bool privacy;
             /* List node for storing AP (owl->ap_list is the head),
              * this field is for interface in AP mode.
              */
             struct list_head ap_list;
         };
     };
+
+    u32 beacon_ie_len;
```
Comment (rickywu0421) [inline vwifi.c:117]: Leave these two fields outside the union. 

I found that there may be some bugs when we use union (inside `struct owl_vif`) to split the fields between STA and HostAP. For example, the line 668 `if (vif->scan_request) {` in `owl_scan()` is dangerous because if the `vif` is operating in HostAP mode, this line will read the memory area of the AP field. Off course we can avoid this by checking the operating mode in the beginning of every function, but sometimes we still forget to do so. 

In conclusion, I think the union make our code prone to bugs. We should no longer use it.

File: vwifi.c [hunk 3]
```diff
@@ -91,12 +106,16 @@ struct owl_vif {
...
+    u8 beacon_ie[IE_MAX_LEN];
```
Comment (rickywu0421) [inline vwifi.c:118]: Leave these two fields outside the union.

I found that there may be some bugs when we use union (inside `struct owl_vif`) to split the fields between STA and HostAP. For example, the line 668 if (`vif->scan_request`) { in `owl_scan()` is dangerous because if the `vif` is operating in HostAP mode, this line will read the memory area of the AP field. Off course we can avoid this by checking the operating mode in the beginning of every function, but sometimes we still forget to do so.

In conclusion, I think the union make our code prone to bugs. We should no longer use it.

Comment (jserv) [reply vwifi.c:118]: > In conclusion, I think the union make our code prone to bugs. We should no longer use it.

Send another pull request?

File: README.md [hunk 4]
```diff
@@ -4,7 +4,7 @@
 such as scanning dummy Wi-Fi networks, connecting, and disconnecting from them.
 It is based on the [cfg80211 subsystem](https://www.kernel.org/doc/html/latest/driver-api/80211/cfg80211.html),
 which works together with FullMAC drivers.
-Currently, it supports station mode and Host AP mode.
+Currently, it supports station mode and Host AP. Also, vwifi is well fitted with the security facilities WPA/WPA2, one can build a wireless environment with `vwifi`, `hostapd` (running on HostAP mode interface), and `wpa_supplicant` (running on station mode interface).
```
Comment (jserv) [inline README.md:7]: Improve the writing.

File: vwifi.c [hunk 5]
```diff
@@ -989,6 +1063,96 @@ static int owl_stop_ap(struct wiphy *wiphy, struct net_device *ndev)
...
 
+static int owl_change_beacon(struct wiphy *wiphy,
+                             struct net_device *ndev,
+                             struct cfg80211_beacon_data *info)
+{
+    struct owl_vif *vif = ndev_get_owl_vif(ndev);
+    int ie_offset = DOT11_MGMT_HDR_LEN + DOT11_BCN_PRB_FIXED_LEN;
+    int head_ie_len, tail_ie_len;
+
+    /* cfg80211 and some upper user-space programs treat IEs as two-part:
```
Comment (jserv) [inline vwifi.c:1074]: Remove "upper" since it is not necessary.

File: vwifi.c (hunk 3 above)
Comment (rickywu0421) [reply vwifi.c:118]: Yes, I will send another PR latter.
